text/1060-tracked-promise.md

```diff
@@ -0,0 +1,494 @@
+```
+
+> [!NOTE]
+> Key behaviors:
+> - if the passed promise is resolved, or a non-promise, we do not await, this allows 
+>   values to not block render (or cause a re-render if they don't need to)
+> - no `@dependentKeyCompat`
+> - promise states are all mutually exclusive
+
+### `trackPromise`
+
+This utility wraps and instruments any promise with reactive state, `TrackedPromise`.  
+
+
+Sample type declaration
+```ts
+export function trackPromise<Value>(
+    existingPromise: Promise<Value> | Value
+): TrackedPromise<Value> {
+    /* ... */
+}
+```
+
+### `TrackedPromise`
+
+This utility is analgous to `new Promise((resolve) => /* ... */)`, but includes the tracking of the underlying promise.
+Additionally `TrackedPromise` must be able to receive an existing promise via `new TrackedPromise(existingPromise);`.
+
+Sample type declaration
+```ts
+export class TrackedPromise<Value = unknown> implements PromiseLike<Value> {
+    constructor(existingPromise: Promise<Value>);
+    constructor(callback: ConstructorParameters<typeof Promise<Value>>[0]);
+    constructor(promiseOrCallback: /* ... */) { /* ... */ }
+
+    // private, tracked, all state is derived from this,
+    // since promises are not allowed to have more than one state. 
+    #state: 
+    | ['PENDING'] 
+    | ['REJECTED', error: unknown] 
+    | ['RESOLVED', value: Value];
+
+    // private, the underlying promise that we're instrumenting
+    #promise: Promise<Value>; 
+
+    // upon success, Value, otherwise null.
+    value: Value | null;
+
+    // upon error, unknown, otherwise null.
+    error: unknown;
+
+    // state helpers
+    isPending: boolean;
+    isResolved: boolean;
+    isRejected: boolean;
+
+    // allows TrackedPromises to be awaited
+    then(/* ... */): PromiseLike</* ... */>
+}
+```
+
+Unlike [TrackedAsyncData][gh-tracked-async-data], all properties are accessible at throughout the the lifecycle of the promise. This is to reduce the pit-of-frustration and eliminate extraneous errors when working with promise-data. While ember-async-data can argue that it doesn't make sense to _even try_ accessing `value` until a promise is resolved, the constraint completely prevents UI patterns where you want value and loading state to be separate, and not part of a single whollistic if/else-if set of blocks. 
+
+For simplicity, these new utilities will not be using `@dependentKeyCompat` to support the `@computed` era of reactivity.  pre-`@tracked` is before ember-source @ 3.13, which is from over 5 years ago, at the time of writing. For the broadest, most supporting libraries we have, 3.28+ is the supported range, and for speed of implementation, these tracked promise utilities can strive for the similar compatibility.
+
+An extra feature that none of the previously mentioned implementations have is the ability to `await` directly. This is made easy by only implementing a `then` method -- and allows a good ergonomic bridge between reactive and non-reactive usages.
```

Comment:
One of the reasons I'm against a built-in reactive promise is actually exactly this. I don't think that something that exposes additional values that the native primitive does not have should also pretend to be the native primitive in this way, because at that point you are a whole different thing.

e.g. this isn't ReactivePromise, this is `AsyncData`

conceptually, a reactive promise would wrap the original promises' methods in one that subscribes to a signal, then finally the original promise and use it that to update the signal once the resolution was complete so that consuming code would repull. Obviously that's far less useful than the state-machine approach taken here and by many of these primitives.

Reply:
Thanks, i agree actually. I've updated the name of the thing so it doesn't imply that it's trying to masquerade as a native promise